perf(ivy): move local references into consts array

crisbeto opened the pull request:
Follow-up from #32798. Moves the local references array into the component def's `consts` in order to make it compress better.

Before:
```
const _c0 = ['foo', ''];

SomeComp.ngComponentDef = defineComponent({
  template: function() {
    element(0, 'div', null, _c0);
  }
});
```

After:
```
SomeComp.ngComponentDef = defineComponent({
  consts: [['foo', '']],
  template: function() {
    element(0, 'div', null, 0);
  }
});
```

crisbeto commented:
I ran some tests on the same app I tested #32798 against and there is a reduction in the gzipped size, but it's less than 1%. I'm guessing it's because template references just aren't as common as attributes.

kara commented on packages/core/src/render3/instructions/container.ts:
```diff
@@ -57,15 +57,15 @@ export function ɵɵcontainer(index: number): void {
  * @param tagName The name of the container element, if applicable
  * @param constsIndex Index of template in the `consts` array.
- * @param localRefs A set of local reference bindings on the element.
+ * @param localRefs Index of the local refenreces in the `consts` array.
```
```suggestion
 * @param localRefs Index of the local references in the `consts` array.
```

kara commented on packages/core/src/render3/instructions/container.ts:
```diff
@@ -74,10 +74,13 @@ export function ɵɵtemplate(
+      tViewConsts === null || constsIndex == null ? null : tViewConsts[constsIndex] as TAttributes);
+  const localRefs = tViewConsts === null || localRefsIndex == null ?
+      null :
+      tViewConsts[localRefsIndex] as string[];
```
Since this pattern keeps repeating, seems like we should have a utility that shares the null checks. e.g.

```ts
getConstant(tViewConsts, attrsIndex) as TAttributes;
  
const localRefs = getConstant(tViewConsts, localRefsIndex) as string[];
```

kara commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -42,25 +42,30 @@ import {registerInitialStylingOnTNode} from './styling';
-
-  ngDevMode && ngDevMode.rendererCreateElement++;
-  ngDevMode && assertDataInRange(lView, index + HEADER_OFFSET);
+  const attrs =
```
These should be `getConstant` too

kara reviewed (approved):
LGTM

kara commented:
[VE presubmit](https://test.corp.google.com/ui#id=OCL:275100099:BASE:275100179:1571257500729:670e847c)

[Ivy presubmit](http://test/OCL:275100099:BASE:275100920:1571257742633:74324151)

crisbeto committed:
perf(ivy): move local references into consts array

Follow-up from #32798. Moves the local references array into the component def's `consts` in order to make it compress better.

Before:
```
const _c0 = ['foo', ''];

SomeComp.ngComponentDef = defineComponent({
  template: function() {
    element(0, 'div', null, _c0);
  }
});
```

After:
```
SomeComp.ngComponentDef = defineComponent({
  consts: [['foo', '']],
  template: function() {
    element(0, 'div', null, 0);
  }
});
```

AndrewKushnir commented:
New presubmits after rebasing:

- [VE Presubmit](http://test/OCL:277370255:BASE:277370304:1572385930477:3e99d39c)
- [Ivy Presubmit](http://test/OCL:277370255:BASE:277371200:1572386185175:1d825af3)

AndrewKushnir commented:
New presubmits (11/1):
- [VE Presubmit](http://test/OCL:278033070:BASE:278033097:1572655157578:fa01b411)
- [Ivy Presubmit](http://test/OCL:278033070:BASE:278033179:1572655199217:3728d1f2)

AndrewKushnir commented:
FYI, VE and Ivy presubmits are successful.

@crisbeto feel free to add "merge" label if it's ready to be merged from your side and no additional approvals/reviews are required.

Thank you.
